File: DSA/Kanikaa_Shree_P_V_2503717610422080/Day_3/Question2.cpp
#include <iostream>
#include <vector>
using namespace std;
int firstO(vector<int>& arr, int target)
 {
    int low = 0, high = arr.size() - 1;
    int first = -1;
    while (low <= high) {
        int mid = low + (high - low) / 2;
        if (arr[mid] == target) 
        {
            first = mid;
            high = mid - 1;
        }
        else if (arr[mid] < target)
        {
            low = mid + 1;
        }
        else 
        {
            high = mid - 1;
        }
    }
    return first;
}
int lastO(vector<int>& arr, int target)
 {
    int low = 0, high = arr.size() - 1;
    int last = -1;
    while (low <= high)
    {
        int mid = low + (high - low) / 2;
        if (arr[mid] == target)
        {
            last = mid;
            low = mid + 1;
        }
        else if (arr[mid] < target)
        {
            low = mid + 1;
        }
        else
        {
            high = mid - 1;
        }
    }
    return last;
}

int main()
{
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }
    int target;
    cout<<"Enter Target value:";
    cin >> target;
    int first = firstO(arr, target);
    if (first == -1)
    {
        cout << 0;
    }
    else
    {
        int last = lastO(arr, target);
        cout << last - first + 1;
    }
    return 0;
}

File: DSA/Sudhanshu_Ranjan_590018435/Day_3/Question2.java
public class Question2 {

    public int search(int[] array, int target) {
        int start = first(array, target);

        if (start == -1) {
            return 0;
        }

        int end = last(array, target);
        return end - start + 1;
    }

    private int first(int[] array, int target) {
        int left = 0;
        int right = array.length - 1;
        int index = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (array[mid] >= target) {
                if (array[mid] == target) {
                    index = mid;
                }
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return index;
    }

    private int last(int[] array, int target) {
        int left = 0;
        int right = array.length - 1;
        int index = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (array[mid] <= target) {
                if (array[mid] == target) {
                    index = mid;
                }
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return index;
    }
}
